0035-ilp-over-http/0035-ilp-over-http.md

```diff
@@ -56,3 +52,115 @@ All ILP Packets MUST be returned with the HTTP status code `200: OK`.
 
 An endpoint MAY return standard HTTP errors, including but not limited to: a malformed or unauthenticated request, rate limiting, or an unresponsive upstream service. Connectors SHOULD either retry the request, if applicable, or relay an ILP Reject packet back to the original sender with an appropriate [Final or Temporary error code](./0027-interledger-protocol-4/0027-interledger-protocol-4#error-codes). Server errors (status codes 500-599) SHOULD be translated into ILP Reject packets with `T00: Temporary Error` codes.
 
+### Authentication
+When authenticating requests between Interledger nodes, it is important to choose an algorithm that maintains an appropriate balance between usability and security, while at the same time maintaining high-performance _and_ interoperability.
+
+In order to find this balance, this document defines two Authentication profiles, each with various trade-offs that should be considered before use:
+
+* `SIMPLE`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret as a [Bearer token](https://tools.ietf.org/html/rfc6750) in all ILP-over-HTTP requests. Peers SHOULD consider this token to be opaque and SHOULD NOT derive any special meaning from the token. 
+
+* `JWT_HS_256`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret in order to _derive_ a Bearer token that conforms to the JSON Web Token (JWT) specification. JWTs generated by this scheme can then be used as a Bearer in all ILP-over-HTTP requests.
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests, but SHOULD support both of the authentication profiles defined in this RFC, and MUST support `JWT_HS_256`.
+ 
+In addition, TLS Client Certificates are RECOMMENDED between peers for additional security. Basic authentication (username and password) is NOT RECOMMENDED for the same reasons that the `SIMPLE` profile is only reccommended for development and testing scenarios only.
+
+#### `SIMPLE` Authentication Profile
+This profile allows two ILP nodes to utilize a previously agreed-upon shared-secret that contains at least 32 bytes (256 bits) of randomly generated data, and is encoded using Base64.
+
+Because tokens in this profile do not inherently contain information about the identity of the caller, requests MUST contain an additional HTTP request-header named `Auth-Principal`.
+ 
+This extra header allows for the identity of the authentication request to be separated from authentication token itself, which reduces computational overhead as well as data-management complexity (e.g., implementations do not need to create data-store indexes using derivations of tokens for lookup purposes).
+
+Depending on the implementation, the value of this header SHOULD represent an `ILP Account Id`, but future implementations MAY make this represent an `ILP Peer Id` or some other value.
```

Comment:
Don't think this paragraph is necessary.

Reply:
Dang, Yeah, I went back and forth on this one. I'd like to remove this paragraph actually, but much of the impetus of this PR is around me trying to peer the Java implementation with other implementations, and needing clarity on certain specifics of ILP-over-HTTP (the Auth systems here are subtly different between Rust, Java, and JS).

If we remove this paragraph, then the meaning of the Principal might be different between implementations (e.g., see [my comments here](https://forum.interledger.org/t/connector-auth-model/517/4?u=sappenin) for one way this could be).

I'm unclear if this is OK or not. For example, in [java-ilp-connector](https://github.com/sappenin/java-ilpv4-connector) I'm going to need to build a Java client that can talk to other connectors. Let's say I'm going to (somehow) work it out with this other Connector that I want to use one of the profiles in this PR. Great. Now let's imagine that my Java Connector wants to use ILP with the other connector over two different accounts (e.g., a USD account and an XRP account). In the Java impl, the current implementation will demand that there be two different HTTP connections, so I'm going to need two different Auth-Principals.

But what if the _other_ Connector in this example actually implements Peer-based Authentication where an ILP-over-HTTP connection can be initiated on a per-peer basis (i.e., One HTTP connection for two peers), and multiple ILP accounts between those peers can be used over the same ILP-over-HTTP connection? In this world, the Java implementation is immediately not going to work because in order to operate on the two accounts in my hypothetical above, the Java Client is going to need a third piece of data -- it will first need the `AuthPrincipal` (Identifies me as `Peer123`); It will then need an Account identifier (USD vs XRP account); and then it will need a credential (e.g., a shared-secret).

Ideally, I'd _love_ not to have to standardize this meaning, but is that reasonable/advisable? Doesn't this meaning need a normative definition?

So, my questions are:

1. How do Connector implementations know what the other party requires means by the `Auth-Principal`? 
1. Does it matter?
1. What happens when two different Connector Implementations decide to define the `Auth-Principal` differently?

Reply:
> How do Connector implementations know what the other party means by the Auth-Principal?

They figure it out before peering

> Does it matter?

Not in a normative sense, I don't think so

> What happens when two different Connector Implementations decide to define the Auth-Principal differently?

They need to figure this out before peering because they need to figure out a bunch of other stuff like extra headers or claims in the JWT to indicate, for example, "the account" when the principal is a peer.

i.e. Let's stick with the simple case of two peers auth with each other and then internally map the principal of that auth to some internal accounting system. For now we will all have an implicit 1:1 mapping of auth principal to account but this may evolve in time at which point we'll need to define how that would work:

1. One auth and additional headers/claims to identify an account, or
2. A different auth principal per account

Reply:
> Not in a normative sense, I don't think so

OK, cool. I like this flexibility and rationale, and [have removed](https://github.com/interledger/rfcs/pull/531/commits/24bca61866b9e7f4004fdd546cfc57f7515a5f7e) any normative definitions of "Authentication Principal" in the spec.

> Let's stick with the simple case of two peers auth with each other and then internally map the principal of that auth to some internal accounting system

Agree. But just to clarify your points, you're saying that even though this is what _most_ of the implementations will do today, we don't need to mandate this...and we don't even want to mention it in the RFC, correct?

I'm open to that path, though I would support a small sentence like this, as a guide to implementors:

`the value of this header SHOULD represent an "ILP Account Id"` `

But I don't feel strongly about including this here, especially given how much we've discussed this in the past. Just wanted to raise this one last time to be sure I've teased out your preference here.

Reply:
> I'm open to that path, though I would support a small sentence like this, as a guide to implementors:

> `the value of this header SHOULD represent an "ILP Account Id" `

I think it is worse with it in for this reason; What IS an `ILP Account Id`? The sentence adds no value without a definition and I don't think this is the spec to define it.

Reply:
> I think it is worse with it in for this reason; What IS an ILP Account Id? The sentence adds no value without a definition and I don't think this is the spec to define it.

Agreed (and I know there's been a lot of back and forth on accounts vs peers), but I do think it's critical that we define what an account is.

The settlement engine spec requires an account identifier shared between itself, the accounting system, and the transport (or at least a way to map that account id to a peer to send messages to, and vice-versa). Also, to support a REST with the account id in the URL, e.g. `POST /account/:id`, we need the account ID to be URL encode-able, which is an additional constraint.

Reply:
> What IS an `ILP Account Id`?

Hmmm....I agree that it's technically not normatively defined anywhere, but the concept of an Account is [pretty-well-covered](https://github.com/interledger/rfcs/blob/master/0001-interledger-architecture/0001-interledger-architecture.md) by RFC, so it shouldn't be a stretch to know what this is. 

In a sentence, it's: "the unique identifier of an Account". 

What's an Account, one might ask? That's easy: go read RFC-0001.

Reply:
> I do think it's critical that we define what an account is (and maybe that's here, maybe that's somewhere else).

I agree with this -- though correct me if I'm wrong @kincaidoneil but I think you're saying, "I do think it's critical that we define what an **accountId** is", right?

Reply:
All that said, I think I'm of the persuasion right now that AccountId is just a String or whaterver two peers want to make it, and so maybe we don't need to normatively define it's _format_ or _structure_. 

But as a concept, I think it should be unambiguous in Interledger to use the term `AccountId` and have people know what that means (as a concept at least).

If that requires a spec somewhere, then I say we add it somewhere.

Reply:
@sappenin Yes (thank you for clarifying!), I meant we should define what an _account id_ is.

> All that said, I think I'm of the persuasion right now that AccountId is just a String or whaterver two peers want to make it, and so maybe we don't need to normatively define it's format or structure.

But I'd say what the account ID represents is very important.

Suppose you have an incoming `peer.settle` packet, and the auth mechanism specifies a peer ID, but not an account ID (one previous topic of discussion). Suppose, if that peer has multiple accounts, when the connector forwards that packet to the settlement engine, which account ID should it specify?

Which is to say, I think it's important we have a way to link incoming packets to the account identifier shared with the settlement engine (which is very related to this spec).

Reply:
> I think it's important we have a way to link incoming packets to the account identifier shared with the settlement engine 

But not as part of the minimum requirements for auth between peers which is what we're trying to define here.

Right now we have an implicit 1:1 association between the entity that has auth'd with me and the account I track for them.

When the possibility exists one day that the entity that auth's with me may have multiple accounts then we'll need to define a way to identify the account but that may never be required so I don't think we need to define it.

Let's start with what we need now and add normative requirements as we discover we need them later.

Reply:
Ok, so summarizing this topic, I see three options. 

I'll list them here with what I see as Pros/Cons for each, then see how everyone feels via Thumbs-up/Thumbs down and/or comments (if you have a "pro" or a "con" that you want added, send it my way and I'll edit this comment).

# Choices for defining the `Auth Principal`
Summarizing this topic, I see three options:

1. **Stay Silent**:  
    1. The RFC is silent as to the meaning of the `sub` claim (aka, the `Auth-Principal`) in the JWTs defined in the `JWT_HS_256` Authentication profile.

1. **Define One JWT Auth Profile**: 
    1. The RFC normatively defines `AccountId` as being the unique identifier of an [Account](https://github.com/interledger/rfcs/blob/master/0032-peering-clearing-settlement/0032-peering-clearing-settlement.md#accounts-and-balances),
    1. For the `JWT_HS_256` auth profile, the RFC normatively defines the `sub` claim of the JWT as representing an `AccountId`. 
    1. If that definition ever needs to change, a new Authentication Profile will be defined.

1. **Define Two JWT Auth Profiles**: 
    1. In addition to defining the above, the RFC _also_  proactively defines a 2nd Authentication Profile (e.g., `JWT_HS_256_PEER`) that is identical to the above except that this 2nd Auth profile instead defines `AccountId` as being the identifier of a [Peer](https://github.com/interledger/rfcs/blob/master/0032-peering-clearing-settlement/0032-peering-clearing-settlement.md#accounts-and-balances).

# Pros/Cons of "Silent"
1. Pro: Allows the most flexibility for current implementations because they can choose any meaning for the Auth Principal that they would like.
1. Con: Creates ambiguity in implementations because implementations can't rely on any particular meaning of the Auth Principal when speaking to another supposedly ILP-over-HTTP compliant system (so any two peers would need to manage this meaning, and potentially code around this meaning, on a case-by-case basis).

# Pros/Cons of "1 Profile"
1. Pro: Clarifies the  meaning of the `sub` claim used in the `JWT_HS_256` profile; allows room for a future profile to be defined; doesn't prematurely define an Auth-Profile that is not currently used or being requested by anyone (i.e., choice 3 below).
1. Con: Precludes the implementation of Peer-wise Auth without an RFC revision to add a 2nd profile (for Peer-wise Auth).

# Pros/Cons of "2 Profiles"
1. Pro: Still clarifies meaning; Defines two different Auth profiles that implementations can choose from, allowing for future possibilities.
1. Con: Specifies Peer-based auth, which is not currently in-use today, and could easily be defined in a future RFC revision if broadly supported, implemented, or required.